### PR TITLE
Add internal hidden rules for testing

### PR DESCRIPTION
**Pull request by zanieb**
Updated implementation of https://github.com/astral-sh/ruff/pull/7369 which was left out in the cold.

This was motivated again following changes in #9691 and #9689 where we could not test the changes without actually deprecating or removing rules.

---

Follow-up to discussion in https://github.com/astral-sh/ruff/pull/7210

Moves integration tests from using rules that are transitively in nursery / preview groups to dedicated test rules that only exist during development. These rules always raise violations (they do not require specific file behavior). The rules are not available in production or in the documentation.

Uses features instead of `cfg(test)` for cross-crate support per https://github.com/rust-lang/cargo/issues/8379

### CODE REVIEW

File: crates/ruff_linter/src/linter.rs [hunk 1]
```diff
@@ -214,6 +216,66 @@ pub fn check_path(
...
+            diagnostics.push(Diagnostic::new(
+                crate::rules::ruff::rules::StableTestRule,
+                ruff_text_size::TextRange::default(),
+            ));
+        }
+
+        // Applying the fix must "resolve" the diagnostic or we will not converge
+        // So only raise the diagnostics on the "first" iteration
+        #[allow(clippy::used_underscore_binding)]
+        if _fix_iteration == 0 {
```
Comment (zanieb) [inline crates/ruff_linter/src/linter.rs:232]: Awkward, but this was really hard to do in the parent calling `check_path`.

File: crates/ruff/tests/integration_test.rs [hunk 2]
```diff
@@ -1169,16 +1171,16 @@ fn check_hints_hidden_unsafe_fixes() {
...
         .pass_stdin("x = {'a': 1, 'a': 1}\n"),
         @r###"
     success: false
     exit_code: 1
     ----- stdout -----
-    -:1:14: F601 Dictionary key literal `'a'` repeated
+    -:1:1: RUF901 [*] Hey this is a stable test rule with a safe fix.
     Found 1 error.
-    No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
+    [*] 1 fixable with the `--fix` option.
```
Comment (zanieb) [inline crates/ruff/tests/integration_test.rs:1183]: Flagging, this is testing the wrong output / selecting the wrong test rule.

File: crates/ruff_linter/src/linter.rs (hunk 1 above)
Comment (BurntSushi) [reply crates/ruff_linter/src/linter.rs:232]: Maybe a comment explaining what `fix_iteration` is? What is the significance of its value? And what is the significance of `0`?

File: crates/ruff/tests/integration_test.rs [hunk 3]
```diff
@@ -1245,47 +1246,41 @@ fn check_shows_unsafe_fixes_with_opt_in() {
...
     ----- stdout -----
-    x = {'a': 1, 'a': 1}
-    print('foo')
+    # safe insertion
 
     ----- stderr -----
-    -:1:14: F601 Dictionary key literal `'a'` repeated
-    Found 2 errors (1 fixed, 1 remaining).
-    No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
+    Found 1 error (1 fixed, 0 remaining).
```
Comment (zanieb) [inline crates/ruff/tests/integration_test.rs:1259]: Flagging, this is a regression in test coverage.

**Review by BurntSushi — APPROVED** [review]
Yeah I think I buy it!

File: crates/ruff_linter/src/linter.rs (hunk 1 above)
Comment (zanieb) [reply crates/ruff_linter/src/linter.rs:232]: I don't think this is right because we will exclude diagnostics that are not fixed... need to investigate

File: crates/ruff_linter/src/linter.rs [hunk 4]
```diff
@@ -521,16 +586,39 @@ pub fn lint_fix<'a>(
...
+        #[cfg(feature = "test-rules")]
+        let filtered_diagnostics = result
+            .data
+            .0
+            .iter()
+            .filter(|diagnostic| !applied_test_fixes.contains(&diagnostic.kind.rule()))
+            .cloned()
+            .collect::<Vec<_>>();
+        #[cfg(feature = "test-rules")]
+        let diagnostics = filtered_diagnostics.as_slice();
```
Comment (zanieb) [inline crates/ruff_linter/src/linter.rs:604]: This is ugly and I do not think it is working correctly — the snapshots are limited to a single fix now but they include diagnostics for the "fixed" test rules

Comment (zanieb) [reply crates/ruff_linter/src/linter.rs:604]: (Because we do not filter these downstream in `result.map`)

File: crates/ruff_linter/src/linter.rs [hunk 5]
```diff
@@ -214,6 +215,39 @@ pub fn check_path(
...
+                    Rule::PreviewTestRule => {
+                        test_rules::PreviewTestRule::diagnostic(&locator, &indexer)
+                    }
+                    _ => unreachable!("All test rules must have an implementation"),
+                };
+                if let Some(diagnostic) = diagnostic {
+                    diagnostics.push(diagnostic);
+                }
+            }
+        }
```
Comment (zanieb) [inline crates/ruff_linter/src/linter.rs:248]: Why is this so awkward Rust... ?

Comment (BurntSushi) [reply crates/ruff_linter/src/linter.rs:248]: I'd probably do

```rust
if !settings.rules.enabled(*test_rule) {
    continue;
}
```

but it otherwise doesn't look too bad?

Comment (zanieb) [reply crates/ruff_linter/src/linter.rs:249]: I don't get why I can't call `::diagnostic` once since I know all of the objects implement `TestRule`.

I have to do some weird dynamic boxing to get to that point?

Comment (BurntSushi) [reply crates/ruff_linter/src/linter.rs]: I don't follow. Can you show me the code you _want_ to write? It looks like you're calling a bunch of different impls of `diagnostic` based on which `Rule` you have? Each line is doing the same function call, but it's on a different type.

Comment (zanieb) [reply crates/ruff_linter/src/linter.rs]: `diagnostic` is provided by the `TestRule` trait which all of these rules implement. For a while I had...

```rust
        for test_rule in TEST_RULES {
            if settings.rules.enabled(*test_rule) {
                if let Some(diagnostic) =
                    implementation_for_test_rule(*test_rule).diagnostic(&locator, &indexer)
                {
                    diagnostics.push(diagnostic);
                }
            }
        }
```

but that required the slightly awkward

```rust
pub(crate) fn implementation_for_test_rule(rule: Rule) -> Box<dyn TestRule>{
    match rule {
        Rule::StableTestRule => Box::new(StableTestRule),
        Rule::StableTestRuleSafeFix => Box::new(StableTestRuleSafeFix),
        Rule::StableTestRuleUnsafeFix => Box::new(StableTestRuleUnsafeFix),
        Rule::StableTestRuleDisplayOnlyFix => Box::new(StableTestRuleDisplayOnlyFix),
        Rule::NurseryTestRule => Box::new(NurseryTestRule),
        Rule::PreviewTestRule => Box::new(PreviewTestRule),
        _ => unreachable!("All test rules must have an implementation"),
    }
}

```

Comment (zanieb) [reply crates/ruff_linter/src/linter.rs]: I was like.. why am I importing `TEST_RULES` and `implementation_for_test_rule` from `test_rules` I should just do something like `iter_test_rules() -> &[Rule, Box<dyn TestRule>]` and that's when I ran into horrible cyclical compiler errors

**Comment by zanieb** [comment]
The ecosystem checks won't succeed until this is main because `RUF9XX` does not exist on the baseline binary.

File: crates/ruff_linter/src/linter.rs (hunk 5 above)
Comment (BurntSushi) [reply crates/ruff_linter/src/linter.rs]: I think any way you slice things here, you need a match statement over a `Rule`. I don't see a way around that because something has to tell the compiler which impl of `TestRule` to use, and _that_ is determined by the `Rule`. Even if you wrote `iter_test_rules() -> &[Rule, Box<dyn TestRule>]`, you'd still need a match statement in there.

I think where you ended up is fine IMO.

File: crates/ruff/Cargo.toml [hunk 6]
```diff
@@ -54,6 +54,8 @@ walkdir = { workspace = true }
 wild = { workspace = true }
 
 [dev-dependencies]
+# Enable test rules during development
+ruff_linter = { path = "../ruff_linter", features = ["clap", "test-rules"] }
```
Comment (charliermarsh) [inline crates/ruff/Cargo.toml:58]: @BurntSushi - Can you double-confirm for me that these won't enable in non-`cargo test` builds?

File: crates/ruff_linter/src/rule_selector.rs [hunk 7]
```diff
@@ -289,6 +289,8 @@ mod schema {
                                 (!prefix.is_empty()).then(|| prefix.to_string())
                             })),
                     )
+                    // Strip the test rules from the schema
+                    .filter(|code| !code.starts_with("RUF9"))
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rule_selector.rs:293]: Wish there were something better here (e.g., what if we could use the rule source, and have a special `lint_source()` called `test`), but might be difficult.

File: python/ruff-ecosystem/ruff_ecosystem/projects.py [hunk 8]
```diff
@@ -206,6 +206,9 @@ def to_ruff_args(self) -> list[str]:
             "check",
             "--no-cache",
             "--exit-zero",
+            # Ignore internal test rules
+            "--ignore",
+            "RUF9",
```
Comment (charliermarsh) [inline python/ruff-ecosystem/ruff_ecosystem/projects.py:211]: Why / how do these get enabled here? Why is this necessary?

**Review by charliermarsh — APPROVED** [review]

File: python/ruff-ecosystem/ruff_ecosystem/projects.py (hunk 8 above)
Comment (charliermarsh) [reply python/ruff-ecosystem/ruff_ecosystem/projects.py:211]: Like how does the feature end up being enabled?

File: crates/ruff_linter/src/linter.rs (hunk 5 above)
Comment (zanieb) [reply crates/ruff_linter/src/linter.rs]: Okay thanks!

File: crates/ruff/Cargo.toml (hunk 6 above)
Comment (zanieb) [reply crates/ruff/Cargo.toml:58]:
```
❯ ./target/release/ruff check example.py --select RUF
❯ ./target/debug/ruff check example.py --select RUF
```

Both do nothing

Comment (zanieb) [reply crates/ruff/Cargo.toml:58]: If I do `cargo build --release --all-features` they are present

```
❯ ./target/release/ruff check example.py --select RUF
example.py:1:1: RUF900 Hey this is a stable test rule.
example.py:1:1: RUF901 [*] Hey this is a stable test rule with a safe fix.
example.py:1:1: RUF902 Hey this is a stable test rule with an unsafe fix.
example.py:1:1: RUF903 Hey this is a stable test rule with a display only fix.
Found 4 errors.
[*] 1 fixable with the `--fix` option (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

but they are not present without `--all-features` in the `release` or `debug` targets.

Comment (zanieb) [reply crates/ruff/Cargo.toml:58]: Idk what features we include in releases, it's buried in maturin?

File: crates/ruff_linter/src/rule_selector.rs (hunk 7 above)
Comment (zanieb) [reply crates/ruff_linter/src/rule_selector.rs:293]: Should we just put them all in a separate lint group e.g. `RFT` for "Ruff test"? Seems safer and easier to filter in general?

Comment (zanieb) [reply crates/ruff_linter/src/rule_selector.rs:293]: Might end up being more work to implement overall though?

Comment (charliermarsh) [reply crates/ruff_linter/src/rule_selector.rs:293]: I'm actually wondering how these even up getting included in the schema, do we run with `--all-features` or something?

Comment (charliermarsh) [reply crates/ruff_linter/src/rule_selector.rs:293]: Anyway, I'm cool with whatever here.

File: crates/ruff/Cargo.toml (hunk 6 above)
Comment (charliermarsh) [reply crates/ruff/Cargo.toml:58]: We don't specify anything, so we get the default features.

Comment (charliermarsh) [reply crates/ruff/Cargo.toml:58]: (Which should omit these.)

Comment (zanieb) [reply crates/ruff/Cargo.toml:58]: Oh great so I think we're good then

File: crates/ruff_linter/src/rule_selector.rs (hunk 7 above)
Comment (zanieb) [reply crates/ruff_linter/src/rule_selector.rs:293]: Huh this raises a preview warning

```
❯ ./target/debug/ruff check example.py --select RUF9
warning: Selection `RUF9` has no effect because the `--preview` flag was not included.
```

but doesn't do anything else.. `RUF8` errors

```
❯ ./target/debug/ruff check example.py --select RUF8
error: invalid value 'RUF8' for '--select <RULE_CODE>'
```

I guess I'll fix that next..

Comment (zanieb) [reply crates/ruff_linter/src/rule_selector.rs:293]: Yeah this is actually confusing... why are they included in the schema in the first place?

Comment (charliermarsh) [reply crates/ruff_linter/src/rule_selector.rs:293]: They shouldn't even _exist_ when we run `cargo dev generate-all`, right?

Comment (zanieb) [reply crates/ruff_linter/src/rule_selector.rs]: Okay this was a bug in the proc macro 🎉 
https://github.com/astral-sh/ruff/pull/9747/commits/8247b8582b4b80900a0ef7e984a035283b4558b6

**Comment by zanieb** [comment]
The schema generation tests use `--all-features` even though the real schema generation does not 😮‍💨

**Comment by charliermarsh** [comment]
Noooooo

File: crates/ruff/Cargo.toml (hunk 6 above)
Comment (BurntSushi) [reply crates/ruff/Cargo.toml:58]: Yeah features enabled in `dev-dependencies` should definitely not cross over into standard builds. But yeah as Zanie mentioned, if you have anything that's using `--all-features`, then they will get enabled in that case.

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -146,7 +149,7 @@ jobs:
       - name: "Run tests"
         shell: bash
         # We can't reject unreferenced snapshots on windows because flake8_executable can't run on windows
-        run: cargo insta test --all --all-features
+        run: cargo insta test --all --exclude ruff_dev --all-features
```
Comment (MichaReiser) [inline .github/workflows/ci.yaml:152]: Can we add another run command that runs the test of `ruff_dev` but with the default features only? We otherwise lose the schema test.
